### PR TITLE
Fix Autocomplete selection display with custom Component

### LINKED ISSUE #4281: AutocompleteInput using a function as optionText shows [object Object] after selecting an item
**What you were expecting:**
When using an optionRenderer function as the optionText for AutocompleteInput, after selecting an item from the list, I hoped to see something useful in the input.

**What happened instead:**
Instead I see [object Object].
image

**Steps to reproduce:**
Code like this should do it:
```

const optionRenderer = choice => {
    return (
         <span>{choice.name}-{choice.id}</span>
    )
};
// ...
<ReferenceInput
    source="post_id"
    reference="posts"
    perPage={15}
    sort={{ field: 'title', order: 'ASC' }}
    fullWidth
>
    <AutocompleteInput
        optionText={optionRenderer}
        options={{ fullWidth: true }}
    />
</ReferenceInput>
```

**Related code:**
See https://codesandbox.io/s/pensive-perlman-vwg90 and navigate to comments and select a comment.

**Other information:**
There seems no other prop besides optionText to render a suggestion once it has been selected. So it kinda makes sense that [object Object] is printed, but I consider it a bug if no other means are available to render the selected item. I suppose I'm missing something.

Somewhat related issues:
- https://github.com/marmelab/react-admin/issues/4092
- https://github.com/marmelab/react-admin/issues/4183
- https://github.com/marmelab/react-admin/issues/3503

**Environment**

* React-admin version: 3.1.2
* Last version that did not exhibit the issue (if applicable): not sure
* React version: 16.12
* Browser: any
* Stack trace (in case of a JS error): -

### PR DESCRIPTION
**Pull request by JulienMattiussi**
Closes #4281

## Issue

When an autocomplete use a custom Component as `optionText`, The selected value can't be display in the filter TextField.

## Solution

When you want to pass a custom component as `optionText` to an Autocomplete, 
you also have to pass a text rendering function returning the filter text to be displayed for your custom component
This `inputText` function must return a `string` which can be similary to the displayed component, or totally different

```jsx
const OptionRenderer = ({ record }) => (
    <span>
        {record.title}-{record.id}
    </span>
);

const inputText = record => `${record.title}-${record.id}`;

<AutocompleteInput
   matchSuggestion={(filterValue, suggestion) => true}
   optionText={<OptionRenderer />}
   inputText={inputText}
   options={{ fullWidth: true }}
/>
```
## Preview

image

## Documentation

image

image

### CODE REVIEW

File: docs/Inputs.md [hunk 1]
```diff
@@ -162,6 +162,28 @@ const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 <AutocompleteInput source="author_id" choices={choices} optionText={optionRenderer} />
 ```
 
+`optionText` also accepts a custom Component. If you do so, you will have to set the inputText function wich determine the input text filter of the current selection:
```
Comment (djhi) [inline docs/Inputs.md]:
```suggestion
`optionText` also accepts a custom Component. If you provide one, you will have to set the `inputText` function wich determine the input text filter of the current selection:
```

**Comment by fzaninotto** [comment]
Please target `next` rather than `master` for this one

File: docs/Inputs.md [hunk 2]
```diff
@@ -162,6 +162,28 @@ const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 <AutocompleteInput source="author_id" choices={choices} optionText={optionRenderer} />
 ```
 
+`optionText` also accepts a custom Component. If you provide one, you will have to set the `inputText` function wich determine the input text filter of the current selection:
```
Comment (fzaninotto) [inline docs/Inputs.md]:
```suggestion
`optionText` also accepts a custom Component. However, as the underlying Autocomplete component requires that the current selection is a string, if you opt for a Component, you must pass a function as the `inputText` prop. This function should return text representation of the current selection:
```

File: docs/Inputs.md [hunk 3]
```diff
@@ -225,8 +247,9 @@ Lastly, would you need to override the props of the suggestions container (a `Po
...
 | `emptyText` | Optional | `string` | `''` | The text to use for the empty element |
 | `matchSuggestion` | Optional | `Function` | - | Required if `optionText` is a React element. Function returning a boolean indicating whether a choice matches the filter. `(filter, choice) => boolean`
-| `optionText` | Optional | <code>string &#124; Function</code> | `name` | Fieldname of record to display in the suggestion item or function which accepts the correct record as argument (`(record)=> {string}`) |
+| `optionText` | Optional | <code>string &#124; Function &#124; Component</code> | `name` | Fieldname of record to display in the suggestion item or function which accepts the correct record as argument (`(record)=> {string}`) |
 | `optionValue` | Optional | `string` | `id` | Fieldname of record containing the value to use as input value  |
+| `inputText` | Optional | <code>Function</code> | `-` | If `optionText` is a custom Component, This function is needed to determine the text displayed filter of the current selection. In Autocomplete, the current selection have to be an editable string and can not be a Component  (`(record)=> {string}`) |
```
Comment (fzaninotto) [inline docs/Inputs.md]:
```suggestion
| `inputText` | Optional | <code>Function</code> | `-` | If `optionText` is a custom Component, this function is needed to determine the text displayed for the current selection. |
```

**Pull request merged**